Rewrite training for Istiod and istio 1.6+ changes (#142)
tetrateio/training: modules/security/rbac/README.md

> # Istio RBAC for Service to Service Communication

Reviewer: ```suggestion
# Istio Authorization for Service to Service Communication
```

Reviewer: can we rename the file too?

> Istio provides every workload with a strong identity - in Kubernetes, the pod's `ServiceAccount` is used to establish mTLS connections between services in the mesh. While establishing mTLS connections, sidecars in the mesh will validate certificates according to [SPIFFE's X.509 SVID spec](https://github.com/spiffe/spiffe/blob/master/standards/X509-SVID.md), which means that after the connection is established we have an authenticated identity of the other party. Then, Istio allows you to write access control policies using those identities to describe which services can communicate.

Reviewer: ```suggestion
Istio provides every workload with a strong identity which is used to establish mTLS connections between services in the mesh. In Kubernetes, the workload's identity is the pod's `ServiceAccount`. While establishing mTLS connections, sidecars in the mesh will validate certificates according to [SPIFFE](https://github.com/spiffe/spiffe/blob/master/standards/X509-SVID.md), which means that after the connection is established we have the _authenticated principal_ of the other party. Then, Istio allows you to write access control policies using those identities to describe which services can communicate.
```

> Note that it means that you have to enforce mTLS between your apps for the Authorization Policies to work.

Reviewer: ```suggestion
Note that it means that you have to enforce mTLS between your apps for Authorization Policies to work.
```

## Setup a deny-all policy

> 

Reviewer: ```suggestion
First, let's start by locking down our application:
```

```yaml
kubectl apply -n hipstershopv1v2 -f - <<EOF
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: ingress-policy
spec:
  selector:
    matchLabels:
      app: istio-ingressgateway
  action: ALLOW
  rules:
  - to:
    - operation:
        methods:
        - GET
---
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: ingress-policy-to-frontend
spec:
  selector:
    matchLabels:
      app: frontend
  action: ALLOW
  rules:
  - to:
    - operation:
        methods:
        - GET
EOF
```

If you browser the Hipstershop store now, you can access the frontend. Sadly, you are still getting an error:

```yaml
Uh, oh!
Something has failed. Below are some details for debugging.

HTTP Status: 500 Internal Server Error
```

If we look at the logs of the `frontend` service, we can see we still have a DENY: `could not retrieve currencies: rpc error: code = PermissionDenied desc = RBAC: access denied`

> 

Reviewer: It might be helpful here to look quickly at the logs on Envoy, and turn up the RBAC logs (`istioctl pc log <pod> --level rbac:debug`).

It would show exactly what policy failed, for example. Not super useful in these exact cases, but can be helpful for users trying to debug more complex policies.

Author: I added a lot more debug info there. Please review and comment. Maybe I put too much informations there... ?

```

This pod can't connect to the `redis-cart` service.

Oh ohhhhhh, Could it be that we broke the production while adding security rules ? The Redis protocol is not HTTP, and does not fall in the `GET`/`POST` category.

As the production is broken, let's quickly add a policy to allow `cartservice` to call the Redis port `6379`:

```yaml
kubectl apply -n hipstershopv1v2 -f - <<EOF
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: redis-policy
spec:
  action: ALLOW
  rules:
  - from:
    - source:
        principals: ["cluster.local/ns/hipstershopv1v2/sa/cartservice"]
    to:
    - operation:
        ports: ["6379"]
EOF
```

Once applied the `cartservice` pod should be back up. If not, you can kill it to recycle it: `kubectl -n hipstershopv1v2 delete pod -l app=cartservice`. Few seconds later your should have a running pod, and the Hipstershop should be back online.

## Cleanup

    For now, lets clean up our Cluster Policy Config so we can carry on with the rest of the lab:

    ```shell
    kubectl -n hipstershopv1v2 delete AuthorizationPolicy deny-all ingress-policy policy-for-hipstershop redis-policy
    ```

## Takeaway

> Once mTLS is enforced, we can use Istio create RBAC policies to tontrol the traffic in the mesh. While this is powerfull, it may also lead to unattended traffic break. Plane, review and tests your rules before applying them in production.

Reviewer: ```suggestion
Once mTLS is enforced, we can use Istio create Authorization Policies to control the traffic in the mesh. While this is powerful, it may also lead to unintended traffic breaks. Plan, review and tests your rules before applying them in production!
```
